src/Mod/Arch/archobjects/reinforcement_custom.py

```diff
@@ -0,0 +1,224 @@
+                "Direction",
+                "Reinforcement",
+                QT_TRANSLATE_NOOP(
+                    "App::Property",
+                    (
+                        "The direction to use to spread the bars. "
+                        "Keep (0,0,0) for automatic direction."
+                    )
+                )
+            )
+        obj.Direction = FreeCAD.Vector(0, 0, 1)
+
+        # length of the rebar distrbution
+        # later if we use a edge too, 0 indicates the use of the edge length
+        if "Distance" not in pl:
+            obj.addProperty(
+                "App::PropertyLength",
+                "Distance",
+                "Reinforcement",
+                QT_TRANSLATE_NOOP(
+                    "App::Property",
+                    (
+                        "The total distance to span the rebars over. "
+                        # "Keep 0 to automatically use the host shape size."
+                    )
+                )
+            )
+        obj.Distance = 400
+        obj.setEditorMode("Amount", 1)
+
+        # concrete cover start
+        if "OffsetStart" not in pl:
+            obj.addProperty(
+                "App::PropertyLength",
+                "OffsetStart",
+                "ArrayOfRebars",
+                QT_TRANSLATE_NOOP(
+                    "App::Property",
+                    (
+                        "The distance between the start of the "
+                        "reinforcement and the first bar (concrete cover)."
+                    )
+                ),
+            )
+        obj.OffsetStart = 0
+
+        # concrete cover end
+        if "OffsetEnd" not in pl:
+            obj.addProperty(
+                "App::PropertyLength",
+                "OffsetEnd",
+                "ArrayOfRebars",
+                QT_TRANSLATE_NOOP(
+                    "App::Property",
+                    (
+                        "The distance between the end of the "
+                        "reinforcement and the last bar (concrete cover)."
+                    )
+                ),
+            )
+        obj.OffsetEnd = 0
+
+        # custom spacing
+        if "CustomSpacing" not in pl:
+            obj.addProperty(
+                "App::PropertyString",
+                "CustomSpacing",
+                "ArrayOfRebars",
+                QT_TRANSLATE_NOOP(
+                    "App::Property",
+                    "The custom spacing of rebars."
+                )
+            )
+
+    def execute(
+        self,
+        obj  # why obj? self is the obj?
+    ):
+
+        # what should be used to access an attribute
+        # self.Attribute
+        # obj.Attribute
+        # same in onChanged()
+
+        if self.clone(obj):
+            return
+        if not obj.BaseRebar:
+            return
+        # should we check for more Attributes?
+
+        from ArchRebar import CustomSpacingPlacement
+        from ArchRebar import strprocessOfCustomSpacing
```

Comment:
Import at the top of the file.

Reply:
BTW: according pep8 from imports are allowed inside the module and make sense IMH

Reply:
They make no sense. All modules should be imported at the top of the module, just like in C you don't `#include <headers>` in the middle of a function. I think the preprocessor technically allows it but nobody would do it.

Adding modules in the middle of a function "hides" the module, and won't immediately report errors in the import. For example, if you introduce errors to `ArchRebar`, you will not notice those errors until you actually use the function at run time. If you import the module at the top of the file, you will find the error when importing `reinforcement_custom`, which then imports `ArchRebar`. Finding the error in this way, just by importing a module is similar to finding an error at compile time in C.

This is the issue we had when restructuring Draft. Some imports where in the middle of classes and functions, but I didn't notice until later. I would have noticed immediately if they had been at the top of the file. See #3261. If `Arch_Panel`, `Arch_Structure`, `Arch_Wall`, and `Arch_Window` had proper imports, I would have found the issue faster. Instead, I had to test all Arch tools to verify there were no more broken tools.

[PEP8 imports](https://www.python.org/dev/peps/pep-0008/#imports)
> Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants.